// @inversifyjs/logger/lib/cjs/logger/modules/LoggerAdapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoggerAdapter = void 0;
const LogLevel_1 = require("../../model/LogLevel");
class LoggerAdapter {
    _loggerOptions;
    #context;
    constructor(context, loggerOptions) {
        this._loggerOptions = this.#parseLoggingOptions(loggerOptions);
        this.#context = context;
    }
    info(message, contextMetadata) {
        this.log(LogLevel_1.LogLevel.INFO, message, contextMetadata);
    }
    http(message, contextMetadata) {
        this.log(LogLevel_1.LogLevel.HTTP, message, contextMetadata);
    }
    silly(message, contextMetadata) {
        this.log(LogLevel_1.LogLevel.SILLY, message, contextMetadata);
    }
    log(logType, message, contextMetadata) {
        if (this._loggerOptions.logTypes.includes(logType)) {
            this.printLog(logType, message, {
                ...contextMetadata,
                context: contextMetadata?.context ?? this.#context,
            });
        }
    }
    error(message, contextMetadata) {
        this.log(LogLevel_1.LogLevel.ERROR, message, contextMetadata);
    }
    warn(message, contextMetadata) {
        this.log(LogLevel_1.LogLevel.WARN, message, contextMetadata);
    }
    debug(message, contextMetadata) {
        this.log(LogLevel_1.LogLevel.DEBUG, message, contextMetadata);
    }
    verbose(message, contextMetadata) {
        this.log(LogLevel_1.LogLevel.VERBOSE, message, contextMetadata);
    }
    #parseLoggingOptions(loggerOptions) {
        return {
            json: loggerOptions?.json ?? false,
            logTypes: loggerOptions?.logTypes ?? [
                LogLevel_1.LogLevel.DEBUG,
                LogLevel_1.LogLevel.ERROR,
                LogLevel_1.LogLevel.HTTP,
                LogLevel_1.LogLevel.INFO,
                LogLevel_1.LogLevel.SILLY,
                LogLevel_1.LogLevel.VERBOSE,
                LogLevel_1.LogLevel.WARN,
            ],
            timestamp: loggerOptions?.timestamp ?? true,
        };
    }
}
exports.LoggerAdapter = LoggerAdapter;
