// @inversifyjs/http-express/lib/cjs/adapter/InversifyExpressHttpAdapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InversifyExpressHttpAdapter = void 0;
const tslib_1 = require("tslib");
const http_core_1 = require("@inversifyjs/http-core");
const cookie_parser_1 = tslib_1.__importDefault(require("cookie-parser"));
const express_1 = tslib_1.__importDefault(require("express"));
const ADAPTER_ID = Symbol.for('@inversifyjs/http-express/InversifyExpressHttpAdapter');
class InversifyExpressHttpAdapter extends http_core_1.InversifyHttpAdapter {
    id = ADAPTER_ID;
    constructor(container, httpAdapterOptions, customApp) {
        super(container, {
            logger: true,
            useCookies: false,
            useJson: true,
            useText: false,
            useUrlEncoded: false,
        }, httpAdapterOptions, [], customApp);
    }
    _buildApp(customApp) {
        return customApp ?? this._buildDefaultExpressApp();
    }
    _buildDefaultExpressApp(customApp) {
        const app = customApp ?? (0, express_1.default)();
        if (this.httpAdapterOptions.useCookies) {
            app.use((0, cookie_parser_1.default)());
        }
        if (this.httpAdapterOptions.useJson) {
            app.use(express_1.default.json({ type: 'application/json' }));
        }
        if (this.httpAdapterOptions.useText) {
            app.use(express_1.default.text({ type: 'text/*' }));
        }
        if (this.httpAdapterOptions.useUrlEncoded) {
            app.use(express_1.default.urlencoded({
                extended: false,
                type: 'application/x-www-form-urlencoded',
            }));
        }
        return app;
    }
    _buildRouter(routerParams) {
        for (const routeParams of routerParams.routeParamsList) {
            const orderedPreHandlerMiddlewareList = [...routeParams.preHandlerMiddlewareList, ...routeParams.guardList];
            const orderedPostHandlerMiddlewareList = routeParams.postHandlerMiddlewareList;
            const normalizedPath = (0, http_core_1.buildNormalizedPath)(`${routerParams.path}${routeParams.path}`);
            this._app[routeParams.requestMethodType](normalizedPath, ...orderedPreHandlerMiddlewareList, routeParams.handler, ...orderedPostHandlerMiddlewareList);
        }
    }
    _replyText(_request, response, value) {
        response.send(value);
    }
    _replyJson(_request, response, value) {
        response.json(value);
    }
    _replyStream(_request, response, value) {
        value.pipe(response);
    }
    _sendBodySeparator(_request, response) {
        response.flushHeaders();
    }
    _setStatus(_request, response, statusCode) {
        response.status(statusCode);
    }
    _setHeader(_request, response, key, value) {
        response.setHeader(key, value);
    }
    _getBody(request, _response, parameterName) {
        return parameterName === undefined
            ? request.body
            : // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
                request.body[parameterName];
    }
    _getParams(request, parameterName) {
        return parameterName === undefined
            ? request.params
            : request.params[parameterName];
    }
    _getQuery(request, parameterName) {
        return parameterName === undefined
            ? request.query
            : request.query[parameterName];
    }
    _getHeaders(request, parameterName) {
        return parameterName === undefined
            ? request.headers
            : request.headers[parameterName];
    }
    _getCookies(request, _response, parameterName) {
        return parameterName === undefined
            ? request.cookies
            : request.cookies[parameterName];
    }
}
exports.InversifyExpressHttpAdapter = InversifyExpressHttpAdapter;
