// @inversifyjs/http-core/lib/cjs/routerExplorer/calculations/buildRouterExplorerControllerMethodMetadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildRouterExplorerControllerMethodMetadata = buildRouterExplorerControllerMethodMetadata;
const framework_core_1 = require("@inversifyjs/framework-core");
const buildErrorTypeToErrorFilterMap_1 = require("./buildErrorTypeToErrorFilterMap");
const getControllerMethodHeaderMetadata_1 = require("./getControllerMethodHeaderMetadata");
const getControllerMethodParameterMetadataList_1 = require("./getControllerMethodParameterMetadataList");
const getControllerMethodStatusCodeMetadata_1 = require("./getControllerMethodStatusCodeMetadata");
const getControllerMethodUseNativeHandlerMetadata_1 = require("./getControllerMethodUseNativeHandlerMetadata");
function buildRouterExplorerControllerMethodMetadata(logger, controllerMetadata, controllerMethodMetadata) {
    const controllerMethodParameterMetadataList = (0, getControllerMethodParameterMetadataList_1.getControllerMethodParameterMetadataList)(controllerMetadata.target, controllerMethodMetadata.methodKey);
    const controllerMethodStatusCode = (0, getControllerMethodStatusCodeMetadata_1.getControllerMethodStatusCodeMetadata)(controllerMetadata.target, controllerMethodMetadata.methodKey);
    const controllerMethodGuardList = [
        ...(0, framework_core_1.getClassGuardList)(controllerMetadata.target),
        ...(0, framework_core_1.getClassMethodGuardList)(controllerMetadata.target, controllerMethodMetadata.methodKey),
    ];
    const controllerMethodInterceptorList = [
        ...(0, framework_core_1.getClassInterceptorList)(controllerMetadata.target),
        ...(0, framework_core_1.getClassMethodInterceptorList)(controllerMetadata.target, controllerMethodMetadata.methodKey),
    ];
    const controllerMiddlewareList = (0, framework_core_1.getClassMiddlewareList)(controllerMetadata.target);
    const controllerMiddlewareOptions = (0, framework_core_1.buildMiddlewareOptionsFromApplyMiddlewareOptions)(controllerMiddlewareList);
    const controllerMethodMiddlewareList = (0, framework_core_1.getClassMethodMiddlewareList)(controllerMetadata.target, controllerMethodMetadata.methodKey);
    const errorTypeToErrorFilterMap = (0, buildErrorTypeToErrorFilterMap_1.buildErrorTypeToErrorFilterMap)(logger, controllerMetadata.target, controllerMethodMetadata.methodKey);
    const controllerMethodMiddlewareOptions = (0, framework_core_1.buildMiddlewareOptionsFromApplyMiddlewareOptions)(controllerMethodMiddlewareList);
    const headerMetadataList = (0, getControllerMethodHeaderMetadata_1.getControllerMethodHeaderMetadata)(controllerMetadata.target, controllerMethodMetadata.methodKey);
    const useNativeHandler = (0, getControllerMethodUseNativeHandlerMetadata_1.getControllerMethodUseNativeHandlerMetadata)(controllerMetadata.target, controllerMethodMetadata.methodKey);
    return {
        errorTypeToErrorFilterMap,
        guardList: controllerMethodGuardList,
        headerMetadataList,
        interceptorList: controllerMethodInterceptorList,
        methodKey: controllerMethodMetadata.methodKey,
        parameterMetadataList: controllerMethodParameterMetadataList,
        path: controllerMethodMetadata.path,
        postHandlerMiddlewareList: [
            ...controllerMiddlewareOptions.postHandlerMiddlewareList,
            ...controllerMethodMiddlewareOptions.postHandlerMiddlewareList,
        ],
        preHandlerMiddlewareList: [
            ...controllerMiddlewareOptions.preHandlerMiddlewareList,
            ...controllerMethodMiddlewareOptions.preHandlerMiddlewareList,
        ],
        requestMethodType: controllerMethodMetadata.requestMethodType,
        statusCode: controllerMethodStatusCode,
        useNativeHandler,
    };
}
